### ISSUE TITLE
Modular LSP lib

### ISSUE THREAD

**Issue by Cyrik** [clojure-lsp/clojure-lsp#807]
I'd like to be able to use the LSP bit of clojure-lsp as a library to build my own lsp server.
Currently, the cli part has a hard dependency on the lib part. This dependency would have to be pulled apart so that it's possible to depend on the lsp server and provide your own handlers.

After trying to pull it a part, I realize we probably need 4 parts:
- common: has shared utility bits and protocols
- lib: mostly stays the same as it is. implements all the clojure features of the lsp
- lsp lib: provides a layer on top of lsp4j and a coercer that handles translating lsp4j objects <-> clojure data
- cli/server: is "just" the entry point that brings it all together. provides the cli and lsp server main functions

@ericdallo what do you think about the split?

**Comment by ericdallo** [comment]
I don't like the name `common`,  it sounds too generic, all the other 3 would use that?

what's the difference between `lsp lib` and `cli/server`?

Would you be interested in using what of those 4 for your server?

**Comment by Cyrik** [comment]
Common could be called `interfaces`, but it's not ideal since there are some utilities in there. I'd probably also pull db in there. I'm happy with any name though. In the end, common is not something that would ever be used alone, it's only needed so that the other three libraries don't need to depend on each other.

LSP lib would be the bit I'd like to use as an LSP author. It would be a library that helps you implement your own LSP. So looking at the current namespaces it would hold `server`, `coercer`, and anything in `common`.
 
Cli/server would hold the whole `main` ns, since it's very Clojure specific and would hold the current `server/run-server!` method.

The cli/server and lib parts would both hold the Clojure specific things, while LSP lib and common could be used by LSP implementers.

The following dependencies would exist: lib -> common, lsp lib -> common, cli/server -> lib & lsp lib. The connection between lib and lsp lib would be provided through a protocol that sits in common.

I hope my ramblings make sense, otherwise I'm happy to expand / change any part of the plan.

**Comment by ericdallo** [comment]
got it, so I know @borkdude is implementing a LSP server as well for other thing, and I know he spent some time on lsp4j as well, maybe we could have this as a separated lib? something like `lsp4clj`? Does that makes sense in the place of `lsp lib` you are mentioning?

**Comment by Cyrik** [comment]
> got it, so I know @borkdude is implementing a LSP server as well for other thing, and I know he spent some time on lsp4j as well, maybe we could have this as a separated lib? something like lsp4clj? Does that makes sense in the place of lsp lib you are mentioning?

Yeah, that's the idea. LSP lib or lsp4clj. I wanted to keep it in the repo and just release it as its own jar or use it as a github dep.

I don't see why all of it depending on common would be a problem, but maybe I'm missing something. It's just a lib in the end. I get the component idea, but I'm not sure there are enough components to make it worth the added complexity. Something like [polylith](https://github.com/polyfy/polylith) might help though.

I'll continue working on the PR. It feels like we want to end up in the same place, so having a working proof of concept will probably help.

**Comment by borkdude** [comment]
I have two LSP servers that I work on, one for clj-kondo and one for zen (a custom EDN language). While working on zen, a fellow developer also pitched the idea of cutting out lsp4j completely since it just wraps a json-rpc protocol. What happens is that we're basically coercing JSON into Java objects into Clojure maps, which is a bit crazy. We could use json-rpc directly.

**Comment by Cyrik** [comment]
@borkdude yeah cutting out lsp4j would be good. That way we might even make the library cljs compatible, which would allow it to run in [vscode.dev](vscode.dev) and similar.
I'd still pull apart clojure-lsp first so we have an interface that doesn't depend on lsp4j and then we can try replacing that bit but still depend on the protocol in `common`. Not sure I described this very well 😅 
What do you think?

**Comment by borkdude** [comment]
Seems good.

**Comment by ericdallo** [comment]
Sounds good @Cyrik, having it as a protocol seems indeed a good start, since replace lsp4j won't be that easy as we depend on multiple service classes, not POJOs only, those services classes handle all the input/output communication + method cancellation, the lsp4clj sounds like other approach we could take later after this refactor

**Comment by Cyrik** [comment]
#805 is now much closer to where I see the end result.
lsp4clj now holds the coercer and most of the server ns. The rest of the work is getting rid of the references to lib and pull out most of the `deftype ClojureLSPServer` into lsp.clj.
Is this the architecture you were thinking of as well? @borkdude @ericdallo

**Comment by Cyrik** [comment]
Ok, so the mucking around work is done! 🥳  (lsp4clj does not depend on any clojure-lsp or lib parts anymore, common protocols only describe standard lsp messages)
Now there are a few design decisions to be made:
- Do we change namespace names to indicate the lib they belong to? (I'd say yes for the lsp4clj and common folders)
- Do we keep IHandler as mostly calling `(fn-name _ some-map)` and fn-name be responsible to pull it apart? (I'd rather pull this apart earlier, so that the IHandler protocol indicates how the values look)
- `clojure-lsp.lsp/LSPServer` is a bit of a mess of values going into the constructor. Is this acceptable?
- ServerCapabilites need to be described somehow (I think a basic map like {:hover-provider true, ...} that I pull apart in `clojure-lsp.lsp/LSPServer` should be fine)
- Extending the server with custom messages means you have to write Java classes (because of attributes). @ericdallo started building a more generic way to do it for `dependencyContents` messages, but seems to have abandoned it. Was there a reason? Do we want to find an easier way to extend the server before we push this out?

@borkdude any opinions? I think once I add a way to configure ServerCapabilities tomorrow this should be usable for clj-kondo and zen, right?

**Comment by borkdude** [comment]
I'm currently too busy to be involved with this and I don't want to block the efforts on this: I trust your and @ericdallo's judgements.

**Comment by ericdallo** [comment]
I still want to take a closer look with attention to that PR refactor @Cyrik, but for now the I can answer the last question about pulling more things, I'd like to pull db caching as mentioned on #658, but I'd leave this for a future refactor for now.

**Comment by Cyrik** [comment]
@ericdallo I updated my list above

**Comment by ericdallo** [comment]
Answering your questions:
> Do we change namespace names to indicate the lib they belong to? (I'd say yes for the lsp4clj and common folders)

Yes, I just realized you were already thinking on that, it's the same I recommended on the PR.

> Do we keep IHandler as mostly calling (fn-name _ some-map) and fn-name be responsible to pull it apart? (I'd rather pull this apart earlier, so that the IHandler protocol indicates how the values look)

I think you meant `ClojureFeatureHandler`? could you elaborate?

> Extending the server with custom messages means you have to write Java classes (because of attributes). @ericdallo started building a more generic way to do it for dependencyContents messages, but seems to have abandoned it. Was there a reason? Do we want to find an easier way to extend the server before we push this out?

No, we never though that code would be that generic/be used outside clojure-lsp, dependencyContent is not related to that, but related to make available a jar content.
Although, I agree making that easier for lsp4clj clients would be nice, we could leave this as a next step somehow.


> Do we pull more stuff from lib, like the transit caching? (agree with @ericdallo, lets do this in a separate PR)

Probably, it's the most part that I'd like to make generic, but we need to think carefully here.

**Comment by Cyrik** [comment]
> I think you meant ClojureFeatureHandler? could you elaborate?

I mean `ILSPFeature` definition has: `(execute-command [this doc])` but now the function that implements it has to pull that apart: `(defn execute-command [{:keys [command arguments]}] ...)`. I'd like `ILSPFeature` to already have the actual params as input, so that it's easier to know what to implement and easier to write a spec for it.
I think this will be easier when we run the incoming messages through the coercion as well instead of only going through java->clj, so I'm happy to wait.

Sidenote: do we know how fast the spec stuff is here? I haven't used it at runtime yet and heard that it's supposed to be a lot slower than malli. So if we do even more with it we should probably first benchmark it a little, unless you already know that it doesn't really matter here.
